cpp/src/parquet/column_writer_test.cc

```diff
@@ -1143,5 +1143,143 @@ TEST(TestColumnWriter, WriteDataPageV2Header) {
+    }
+    writer_properties_ = builder.build();
+    metadata_ = ColumnChunkMetaDataBuilder::Make(writer_properties_,
+                                                 schema_descriptor_->Column(0));
+
+    std::unique_ptr<PageWriter> pager = PageWriter::Open(
+        sink_, compression, Codec::UseDefaultCompressionLevel(), metadata_.get(),
+        /* row_group_ordinal */ -1, /* column_chunk_ordinal*/ -1,
+        ::arrow::default_memory_pool(), /* buffered_row_group */ buffered,
+        /* header_encryptor */ NULLPTR, /* data_encryptor */ NULLPTR,
+        /* enable_checksum */ false);
+    return std::static_pointer_cast<parquet::Int32Writer>(
+        ColumnWriter::Make(metadata_.get(), std::move(pager), writer_properties_.get()));
+  }
+
+  std::shared_ptr<::arrow::io::BufferOutputStream> sink_;
+  std::shared_ptr<GroupNode> node_;
+  std::unique_ptr<SchemaDescriptor> schema_descriptor_;
+
+  std::shared_ptr<WriterProperties> writer_properties_;
+  std::unique_ptr<ColumnChunkMetaDataBuilder> metadata_;
+};
+
+TEST_F(TestSizeEstimated, NonBuffered) {
+  auto required_writer = this->BuildWriter(Compression::UNCOMPRESSED, false);
+
+  // Write half page
+  for (int32_t i = 0; i < 50; i++) {
+    required_writer->WriteBatch(1, nullptr, nullptr, &i);
+  }
+  // Page flushed, check size
+  EXPECT_EQ(0, required_writer->total_bytes_written());
+  EXPECT_EQ(0, required_writer->total_compressed_bytes());  // unbuffered
+  EXPECT_EQ(0, required_writer->total_compressed_bytes_written());
+  for (int32_t i = 0; i < 50; i++) {
+    required_writer->WriteBatch(1, nullptr, nullptr, &i);
+  }
+  // Page flushed, check size
+  EXPECT_LT(400, required_writer->total_bytes_written());
+  EXPECT_EQ(0, required_writer->total_compressed_bytes());
+  EXPECT_LT(400, required_writer->total_compressed_bytes_written());
+}
+
+TEST_F(TestSizeEstimated, Buffered) {
+  auto required_writer = this->BuildWriter(Compression::UNCOMPRESSED, true);
+
+  // Write half page
+  for (int32_t i = 0; i < 50; i++) {
+    required_writer->WriteBatch(1, nullptr, nullptr, &i);
+  }
+  // Page flushed, check size
+  EXPECT_EQ(0, required_writer->total_bytes_written());
+  EXPECT_EQ(0, required_writer->total_compressed_bytes());  // buffered
+  EXPECT_EQ(0, required_writer->total_compressed_bytes_written());
+  for (int32_t i = 0; i < 50; i++) {
+    required_writer->WriteBatch(1, nullptr, nullptr, &i);
+  }
+  // Page flushed, check size
+  EXPECT_LT(400, required_writer->total_bytes_written());
+  EXPECT_EQ(0, required_writer->total_compressed_bytes());
+  EXPECT_LT(400, required_writer->total_compressed_bytes_written());
+}
+
+TEST_F(TestSizeEstimated, NonBufferedDictionary) {
+  auto required_writer = this->BuildWriter(Compression::UNCOMPRESSED, false, true);
+
+  // Write half page
+  // for dict, keep value equal
+  int32_t dict_value = 1;
+  for (int32_t i = 0; i < 50; i++) {
+    required_writer->WriteBatch(1, nullptr, nullptr, &dict_value);
+  }
+  // Page flushed, check size
```

Comment:
```suggestion
```
I don't think the page is flushed yet based on the comment below.

Reply:
Yes, nice catch